biz.aQute.bndlib/src/aQute/bnd/http/HttpClient.java

```diff
@@ -247,29 +247,35 @@ public TaggedData send0(final HttpRequest< ? > request) throws Exception {
 
 		configureHttpConnection(request.verb, hcon);
 
-		final ProgressPlugin.Task task = getTask(request);
+		final List<Task> tasks = getTasks(request);
 		try {
 
 			TaggedData td = connectWithProxy(proxy, new Callable<TaggedData>() {
 				@Override
 				public TaggedData call() throws Exception {
-					return doConnect(request.upload, request.download, con, hcon, request, task);
+					return doConnect(request.upload, request.download, con, hcon, request, tasks);
 				}
 			});
 			reporter.trace("result %s", td);
 			return td;
 		} catch (Throwable t) {
-			task.done("Failed " + t.getMessage(), t);
+			for (ProgressPlugin.Task task : tasks) {
+				task.done("Failed " + t.getMessage(), t);
+			}
 			throw t;
 		}
 	}
 
-	ProgressPlugin.Task getTask(final HttpRequest< ? > request) {
-		final ProgressPlugin.Task task;
-		if (progress != null) {
-			task = progress.startTask("Download " + request.url, 100);
+	List<Task> getTasks(final HttpRequest< ? > request) {
```

Comment:
One way to minimize the broad effects of supporting multiple ProgressPlugins would be to not return a List<Task> here but instead a single Task which multiplexed the list of tasks. Then most of the other changes become unnecessary.

So if there is zero ProgressPlugins, return the noop Task. If there is only one ProgressPlugin, return the Task for it. If there are more than one ProgressPlugin, return a MultiplexerTask whose methods call all the actual Tasks methods.